Refs #18558 - Add advanced sync options

jlsherrill opened the pull request:
This adds two new sync options to the sync action:
* skip_metadata_check - pulp's version of 'force_full'
* validate_contents - uses on_demand, repository download action
   with verify_all_units to verify checksums and redownload corrupt
   or missing packages

ehelms commented on app/controllers/katello/api/v2/repositories_controller.rb:
```diff
@@ -181,6 +181,7 @@ def show
     param :id, :identifier, :required => true, :desc => N_("repository ID")
     param :source_url, String, :desc => N_("temporarily override feed URL for sync"), :required => false
     param :incremental, :bool, :desc => N_("perform an incremental import"), :required => false
+    param :force_full, :bool, :desc => N_("Force sync even if no upstream changes are detected"), :required => false
```
We don't use the term 'full' in our descriptions so I wonder if we need it in the parameter name. `force` or `full_sync` would seem to cover it depending on how we want to position the parameter. With 'force' seeming to be closer given we are pushing it to do behaviors that are not natural.

jlsherrill replied:
This comes from the pulp option (https://docs.pulpproject.org/en/2.13/nightly/user-guide/release-notes/2.10.x.html).  I was attempting to keep these the same for simplicity, but if you really dislike it, we can change it.

ehelms replied:
Is this another option that only applies to yum repositories?

jlsherrill replied:
Indeedily doodily neighbor.  I guess i could mention that in the api docs.

ehelms commented on app/controllers/katello/api/v2/repositories_controller.rb:
```diff
@@ -190,8 +191,12 @@ def sync
 
+      sync_options = {
+        :force_full => ::Foreman::Cast.to_bool(params[:force_full])
+      };
```
I tend to agree with the hound on this one :)

jlsherrill commented:
Updated/commented

ehelms commented:
Is it only supported for yum repositories due to Pulp or our own restriction?

jlsherrill commented:
The actual command runs without error for all types, the optimizations that force_full are skipping are not present on any repos but yum repos.    So you would never need to run it on any repo but yum, but wouldn't harm anything

jlsherrill commented:
@ehelms i removed the UI bits for now (for a follow PR), and went ahead and added the 'verify' options.  I also changed the param for 'force_full' as it became more evident as I added the validate that there wasn't much value in trying to preserve pulp naming

ehelms commented on app/controllers/katello/api/v2/repositories_controller.rb:
```diff
@@ -181,7 +181,16 @@ def show
     param :source_url, String, :desc => N_("temporarily override feed URL for sync"), :required => false
     param :incremental, :bool, :desc => N_("perform an incremental import"), :required => false
+    param :skip_metadata_check, :bool, :desc => N_("Force sync even if no upstream changes are detected. Only used with yum repositories."), :required => false
+    param :validate_contents, :bool, :desc => N_("Force a sync and validate the checksums of all contnet. Only used with yum repositories."), :required => false
```
s/contnet/content

ehelms commented on app/lib/actions/katello/repository/sync.rb:
```diff
@@ -16,27 +16,37 @@ class Sync < Actions::EntryAction
+
+          fail _("Unable to sync repo. This repository does not have a feed url.") if repo.url.blank? && source_url.blank?
+          fail _("Cannot validate contents on non-yum repositories.") if validate_contents && !repo.yum?
+          fail _("Cannot skip metadata check on non-yum repositories.") if skip_metadata_check && !repo.yum?
```
Hrmm, will the controller propagate these to the user with a proper HTTP code? I can test, but I have this feeling these would result in 500s which would be overkill.

ehelms commented on app/lib/actions/pulp/repository/sync.rb:
```diff
@@ -27,11 +28,12 @@ def invoke_external_task
 
+            if input[:options]
+              sync_options.merge(input[:options])
+            end
```
Could do this as a one liner if statement.

jlsherrill commented:
@ehelms updated
